scanpy/external/pp/_hashsolo.py

```diff
@@ -0,0 +1,387 @@
+    """
+    priors = np.array(priors)
+    log_likelihoods_for_each_hypothesis, _, _ = _calculate_log_likelihoods(
+        data, number_of_noise_barcodes
+    )
+    probs_hypotheses = (
+        np.exp(log_likelihoods_for_each_hypothesis)
+        * priors
+        / np.sum(
+            np.multiply(np.exp(log_likelihoods_for_each_hypothesis), priors), axis=1
+        )[:, None]
+    )
+    most_likely_hypothesis = np.argmax(probs_hypotheses, axis=1)
+    return {
+        "most_likely_hypothesis": most_likely_hypothesis,
+        "probs_hypotheses": probs_hypotheses,
+        "log_likelihoods_for_each_hypothesis": log_likelihoods_for_each_hypothesis,
+    }
+
+
+def hashsolo(
+    cell_hashing_adata: anndata.AnnData,
+    cell_hashing_columns: list,
+    priors: list = [0.01, 0.8, 0.19],
+    pre_existing_clusters: str = None,
+    number_of_noise_barcodes: int = None,
+    inplace: bool = True,
+):
+    """Probabilistic demultiplexing of cell hashing data using HashSolo [Bernstein20]_.
+
+    .. note::
+        More information and bug reports `here <https://github.com/calico/solo>`__.
+    Parameters
+    ----------
+    cell_hashing_adata : anndata.AnnData
+        Anndata object with cell hashes in .obs columns
+    cell_hashing_columns : list,
+        list specifying which columns in cell_hashing_adata.obs
+        are cell hashing counts
+    priors : list,
+        a list of your prior for each hypothesis
+        first element is your prior for the negative hypothesis
+        second element is your prior for the singlet hypothesis
+        third element is your prior for the doublet hypothesis
+        We use [0.01, 0.8, 0.19] by default because we assume the barcodes
+        in your cell hashing matrix are those cells which have passed QC
+        in the transcriptome space, e.g. UMI counts, pct mito reads, etc.
+    pre_existing_clusters : str
+        column in cell_hashing_adata.obs for how to break up demultiplexing
+        for example leiden or cell types, not batches though
+    number_of_noise_barcodes : int,
+        Use this if you wish change the number of barcodes used to create the
+        noise distribution. The default is number of cell hashes - 2.
+    inplace : bool
+        To do operation in place
+    Returns
+    -------
+    cell_hashing_adata : AnnData
+        if inplace is False returns AnnData with demultiplexing results
+        in .obs attribute otherwise does is in place
+
+    Examples
+    -------
+    >>> import anndata
+    >>> import scanpy.external as sce
+    >>> cell_hashing_data = anndata.read("cell_hashing_counts.h5ad")
+    >>> sce.pp.hashsolo(cell_hashing_data)
+    >>> cell_hashing_data.obs.head()
+    """
+    print(
+        "Please cite HashSolo paper: \nhttps://www.cell.com/cell-systems/fulltext/S2405-4712(20)30195-2"
+    )
+
+    data = cell_hashing_adata.obs[cell_hashing_columns].values
+    if np.any(data < 0):
+        raise ValueError("Cell hashing counts must be non-negative")
```

Comment:
You should be able to use the utility to check for count data 

https://github.com/theislab/scanpy/blob/62bb6431c1b91f56141e05effc09cf21a73665f4/scanpy/_utils.py#L460-L475

Reply:
Ah good to know!